cmd/stream/net_stream.go

```
@@ -0,0 +1,120 @@
+		errs:     make(chan error),
+	}
+}
+
+// In returns a read-only channel of Commands which have been received from the
+// client.
+func (n *NetStream) In() <-chan Command { return n.in }
+
+// Status returns a channel which, when written to, will write out the given
+// Status command.
+func (n *NetStream) Status() chan<- *Status { return n.statuses }
+
+// Errs returns a read-only channel of errors encountered during the Listen
+// operation.
+func (n *NetStream) Errs() <-chan error { return n.errs }
+
+// Close closes the Listen routine. Calling this function blocks until the
+// Listen routine has entered a closing state. Should this function be called
+// while a parse or send operation is taking place, then that operation will
+// finish before the close operation takes place immediately afterwords.
+//
+// Closed NetStreams may be reopened in the future.
+func (n *NetStream) Close() { n.closer <- struct{}{} }
+
+// Listen loops infinitely, managing the incoming and outgoing channel of chunks
+// on the chunk stream shared between the server and client.
+//
+// Listen has three main goals:
+//  - Parse incoming chunks, returning errors when they are unparsable.
+//  - Serialize outgoing `onStatus` commands, returning an error when they are
+//    either unserializable, or unwriteable.
+//  - Respond to the `Close()` operation.
+//
+// Listen runs within its own goroutine, and any errors encountered while
+// running are sent over the internal errs channel, accessible from the `Errs()`
+// function.
+func (n *NetStream) Listen() {
+	for {
+		select {
+		case chunk := <-n.chunks:
+			cmd, err := n.parser.Parse(bytes.NewReader(chunk.Data))
+			if err != nil {
+				n.errs <- err
+				continue
+			}
+
+			n.in <- cmd
+		case st := <-n.statuses:
+			c, err := st.AsChunk()
+			if err != nil {
+				n.errs <- err
+				continue
+			}
+
+			if err = n.writer.Write(c); err != nil {
+				n.errs <- err
+			}
+		case <-n.closer:
+			return
+		}
```

General thoughts on this:
 - I think the the Listen() channel should call close() on the closer channel when the loop ends, so that subsequent attempts to write to the closer will panic (easy to fix) instead of deadlock ("well I guess I'm leaking goroutines")
 - Worker loops should close their output channels when they end, so that other routines that pull from that channel. This lets the system close in a cascading, graceful manner rather than firing off a `.Close()` to everything and hoping that nothing important gets lost or locks.

Don't know if this is something you want to implement here, but I've started to follow these and found them to work quite nicely for less error-prone teardowns.